src/OpenTelemetry/Trace/TracerProviderSdk.cs

```diff
@@ -210,56 +92,68 @@ internal sealed class TracerProviderSdk : TracerProvider
-                    // or not.
-                    listener.ShouldListenTo = (activitySource) =>
-                        this.supportLegacyActivity ?
-                        string.IsNullOrEmpty(activitySource.Name) || regex.IsMatch(activitySource.Name) :
-                        regex.IsMatch(activitySource.Name);
-                }
-                else
-                {
-                    var activitySources = new HashSet<string>(sources, StringComparer.OrdinalIgnoreCase);
+            ActivitySource.AddActivityListener(listener);
+            this.listener = listener;
+        }
 
-                    if (this.supportLegacyActivity)
-                    {
-                        activitySources.Add(string.Empty);
-                    }
+        internal Resource Resource { get; }
 
-                    // Function which takes ActivitySource and returns true/false to indicate if it should be subscribed to
-                    // or not.
-                    listener.ShouldListenTo = (activitySource) => activitySources.Contains(activitySource.Name);
-                }
+        internal ConcurrentBag<object> Instrumentations => this.instrumentations;
+
+        internal BaseProcessor<Activity> Processor => this.processor;
+
+        internal Sampler Sampler => this.sampler;
+
+        public override TracerProvider AddSource(string source)
+        {
+            Guard.ThrowIfNullOrWhitespace(source);
+
+            if (!this.sources.Contains(source))
+            {
+                this.sources.Add(source);
+
+                SetListenerFilter(this.listener, this.sources, this.supportLegacyActivity);
             }
-            else
+
+            return this;
+        }
+
+        public override TracerProvider AddLegacySource(string source)
+        {
+            Guard.ThrowIfNullOrWhitespace(source);
+
+            if (!this.legacySources.Contains(source))
             {
-                if (this.supportLegacyActivity)
-                {
-                    listener.ShouldListenTo = (activitySource) => string.IsNullOrEmpty(activitySource.Name);
-                }
+                this.legacySources.Add(source);
+
+                this.supportLegacyActivity = true;
+                this.ConfigureLegacySourceListener(this.listener, this.legacySources);
             }
 
-            ActivitySource.AddActivityListener(listener);
-            this.listener = listener;
+            return this;
         }
 
-        internal Resource Resource { get; }
+        public override TracerProvider AddInstrumentation(object instrumentation)
+        {
+            Guard.ThrowIfNull(instrumentation);
 
-        internal List<object> Instrumentations => this.instrumentations;
+            this.instrumentations.Add(instrumentation);
 
-        internal BaseProcessor<Activity> Processor => this.processor;
+            return this;
+        }
 
-        internal Sampler Sampler => this.sampler;
+        public override TracerProvider UpdateListener()
```

Comment:
Hack to update internal listeners, due `ActivityListener.ShouldListenTo` is not reactive.